AssertionError: getNewNotificationId() called with no free notification ID

Crash reported via Google Play.

App version: 5.111
Android versions: 6.0, 5.1

```
java.lang.AssertionError: getNewNotificationId() called with no free notification ID
	at com.fsck.k9.notification.NotificationData.getNewNotificationId(NotificationData.java:77)
	at com.fsck.k9.notification.NotificationData.addNotificationContent(NotificationData.java:46)
	at com.fsck.k9.notification.NewMailNotifications.addNewMailNotification(NewMailNotifications.java:56)
	at com.fsck.k9.notification.NotificationController.addNewMailNotification(NotificationController.java:105)
	at com.fsck.k9.controller.MessagingController$12.messageFinished(MessagingController.java:1435)
	at com.fsck.k9.mail.store.pop3.Pop3Store$Pop3Folder.fetch(Pop3Store.java:836)
	at com.fsck.k9.controller.MessagingController.downloadSmallMessages(MessagingController.java:1393)
	at com.fsck.k9.controller.MessagingController.downloadMessages(MessagingController.java:1196)
	at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:953)
	at com.fsck.k9.controller.MessagingController$9.run(MessagingController.java:739)
	at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:204)
	at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:115)
	at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:173)
	at java.lang.Thread.run(Thread.java:818)
```

```
java.lang.AssertionError: getNewNotificationId() called with no free notification ID
	at com.fsck.k9.notification.NotificationData.getNewNotificationId(NotificationData.java:77)
	at com.fsck.k9.notification.NotificationData.addNotificationContent(NotificationData.java:46)
	at com.fsck.k9.notification.NewMailNotifications.addNewMailNotification(NewMailNotifications.java:56)
	at com.fsck.k9.notification.NotificationController.addNewMailNotification(NotificationController.java:105)
	at com.fsck.k9.controller.MessagingController$12.messageFinished(MessagingController.java:1435)
	at com.fsck.k9.mail.store.imap.ImapFolder.fetch(ImapFolder.java:770)
	at com.fsck.k9.controller.MessagingController.downloadSmallMessages(MessagingController.java:1393)
	at com.fsck.k9.controller.MessagingController.downloadMessages(MessagingController.java:1196)
	at com.fsck.k9.controller.MessagingController.access$3200(MessagingController.java:115)
	at com.fsck.k9.controller.MessagingController$40.run(MessagingController.java:4440)
	at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:204)
	at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:115)
	at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:173)
	at java.lang.Thread.run(Thread.java:818)
```

```
java.lang.AssertionError: getNewNotificationId() called with no free notification ID
	at com.fsck.k9.notification.NotificationData.getNewNotificationId(NotificationData.java:77)
	at com.fsck.k9.notification.NotificationData.addNotificationContent(NotificationData.java:46)
	at com.fsck.k9.notification.NewMailNotifications.addNewMailNotification(NewMailNotifications.java:56)
	at com.fsck.k9.notification.NotificationController.addNewMailNotification(NotificationController.java:105)
	at com.fsck.k9.controller.MessagingController.downloadLargeMessages(MessagingController.java:1571)
	at com.fsck.k9.controller.MessagingController.downloadMessages(MessagingController.java:1203)
	at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:953)
	at com.fsck.k9.controller.MessagingController$9.run(MessagingController.java:739)
	at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:204)
	at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:115)
	at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:173)
	at java.lang.Thread.run(Thread.java:818)
```

--- a/k9mail/src/main/java/com/fsck/k9/notification/NotificationData.java
+++ b/k9mail/src/main/java/com/fsck/k9/notification/NotificationData.java
@@ -85,6 +85,10 @@ class NotificationData {
         notificationIdsInUse.put(notificationId, true);
     }
 
+    private void markNotificationIdAsFree(int notificationId) {
+        notificationIdsInUse.delete(notificationId);
+    }
+
     NotificationHolder createNotificationHolder(int notificationId, NotificationContent content) {
         return new NotificationHolder(notificationId, content);
     }
@@ -165,7 +169,7 @@ class NotificationData {
         activeNotifications.remove(holder);
 
         int notificationId = holder.notificationId;
-        notificationIdsInUse.delete(notificationId);
+        markNotificationIdAsFree(notificationId);
 
         if (!additionalNotifications.isEmpty()) {
             NotificationContent newContent = additionalNotifications.removeFirst();

--- a/k9mail/src/main/java/com/fsck/k9/notification/NotificationData.java
+++ b/k9mail/src/main/java/com/fsck/k9/notification/NotificationData.java
@@ -165,6 +165,8 @@ class NotificationData {
         activeNotifications.remove(holder);
 
         int notificationId = holder.notificationId;
+        notificationIdsInUse.delete(notificationId);
+
         if (!additionalNotifications.isEmpty()) {
             NotificationContent newContent = additionalNotifications.removeFirst();
             NotificationHolder replacement = createNotificationHolder(notificationId, newContent);

--- a/k9mail/src/test/java/com/fsck/k9/notification/NotificationDataTest.java
+++ b/k9mail/src/test/java/com/fsck/k9/notification/NotificationDataTest.java
@@ -52,8 +52,7 @@ public class NotificationDataTest {
 
     @Test
     public void testAddNotificationContentWithReplacingNotification() throws Exception {
-        NotificationContent content = createNotificationContent("1");
-        notificationData.addNotificationContent(content);
+        notificationData.addNotificationContent(createNotificationContent("1"));
         notificationData.addNotificationContent(createNotificationContent("2"));
         notificationData.addNotificationContent(createNotificationContent("3"));
         notificationData.addNotificationContent(createNotificationContent("4"));
@@ -105,6 +104,15 @@ public class NotificationDataTest {
         assertNotNull(holder);
         assertEquals(NotificationIds.getNewMailStackedNotificationId(account, 1), holder.notificationId);
         assertEquals(content, holder.content);
+    }
+
+    @Test
+    public void testRemoveDoesntLeakNotificationIds() {
+        for (int i = 1; i <= NotificationData.MAX_NUMBER_OF_STACKED_NOTIFICATIONS+1; i++) {
+            NotificationContent content = createNotificationContent(""+i);
+            notificationData.addNotificationContent(content);
+            notificationData.removeNotificationForMessage(content.messageReference);
+        }
     }
 
     @Test

--- a/k9mail/src/test/java/com/fsck/k9/notification/NotificationDataTest.java
+++ b/k9mail/src/test/java/com/fsck/k9/notification/NotificationDataTest.java
@@ -5,6 +5,7 @@ import java.util.List;
 
 import com.fsck.k9.Account;
 import com.fsck.k9.activity.MessageReference;
+
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -107,9 +108,9 @@ public class NotificationDataTest {
     }
 
     @Test
-    public void testRemoveDoesntLeakNotificationIds() {
-        for (int i = 1; i <= NotificationData.MAX_NUMBER_OF_STACKED_NOTIFICATIONS+1; i++) {
-            NotificationContent content = createNotificationContent(""+i);
+    public void testRemoveDoesNotLeakNotificationIds() {
+        for (int i = 1; i <= NotificationData.MAX_NUMBER_OF_STACKED_NOTIFICATIONS + 1; i++) {
+            NotificationContent content = createNotificationContent("" + i);
             notificationData.addNotificationContent(content);
             notificationData.removeNotificationForMessage(content.messageReference);
         }
@@ -275,7 +276,7 @@ public class NotificationDataTest {
     private MessageReference createMessageReference(String uid) {
         return new MessageReference(ACCOUNT_UUID, FOLDER_NAME, uid, null);
     }
-    
+
     private NotificationContent createNotificationContent(String uid) {
         MessageReference messageReference = createMessageReference(uid);
         return createNotificationContent(messageReference);

